src/lib.rs

```diff
@@ -481,67 +755,163 @@ impl<I: Interchange> Responder<I> {
+                .store(State::Responded as u8, Ordering::Release);
             Ok(())
         } else {
-            Err(())
+            Err(Error)
         }
     }
 }
 
-impl<I: Interchange> Responder<I>
+impl<'i, Q, A> Responder<'i, Q, A>
 where
-    I::RESPONSE: Default,
+    A: Default,
 {
-    /// If there is a request waiting that no longer needs to
-    /// be accessed, may build response into the returned value.
-    pub fn response_mut(&mut self) -> Option<&mut I::RESPONSE> {
+    /// Initialize a response with its default values and mutates it with `f`
+    ///
+    /// This is usefull to build large structures in-place
+    pub fn with_response_mut<R>(&mut self, f: impl FnOnce(&mut A) -> R) -> Result<R, Error> {
         if self.transition(State::Requested, State::BuildingResponse)
             || self.transition(State::BuildingResponse, State::BuildingResponse)
         {
-            unsafe {
-                if !(*self.interchange()).is_response_state() {
-                    *self.interchange_mut() = I::from_rp(&I::RESPONSE::default());
-                }
-
-                Some((*self.interchange_mut()).rp_mut())
-            }
+            let res = unsafe {
+                self.with_data_mut(|i| {
+                    if !i.is_response_state() {
+                        *i = Message::from_rp(A::default());
+                    }
+                    f(i.rp_mut())
+                })
+            };
+            Ok(res)
         } else {
-            None
+            Err(Error)
         }
     }
 
-    /// Send a response that was already placed in the interchange using `response_mut`.
-    pub fn send_response(&mut self) -> Result<(), ()> {
-        if State::BuildingResponse == self.state.load(Ordering::Acquire) {
+    /// Initialize a response with its default values and and return a mutable reference to it
+    ///
+    /// This is usefull to build large structures in-place
+    // Safety: We cannot test this with loom efficiently, but given that `with_response_mut` is tested,
+    // this is likely correct
+    #[cfg(not(loom))]
+    pub fn response(&mut self) -> Result<&mut A, Error> {
+        if self.transition(State::Requested, State::BuildingResponse)
+            || self.transition(State::BuildingResponse, State::BuildingResponse)
+        {
             unsafe {
-                *self.interchange_mut() = I::from_rp(self.response_mut().unwrap());
-            }
-            if self.transition(State::BuildingResponse, State::Responded) {
-                Ok(())
-            } else {
-                Err(())
+                self.with_data_mut(|i| {
+                    if !i.is_response_state() {
+                        *i = Message::from_rp(A::default());
+                    }
+                })
             }
+            Ok(unsafe { self.data_mut().rp_mut() })
+        } else {
+            Err(Error)
+        }
+    }
+
+    /// Send a response that was already placed in the channel using `response_mut` or
```

Comment:
s/response_mut/response/

Reply:
I've renamed the method to `response_mut` for consistency with the requester